### DISCUSSION TITLE
Prisma + Supabase workflow and env vars in Github Actions & Google Cloud Build / Run

### DISCUSSION THREAD

**Discussion by sauntimo** [prisma/prisma#11287] [Q&A] [answered]
I am in the process of migrating a node/typescript API from firebase to supabase + prisma. I have the following builds I need to manage;

Two are for PR checks, both of these fire every time a commit is pushed to a branch on github
- a Github workflow which builds the code and runs a suite of tests
- a Google Cloud Build trigger which builds the code to see if it builds but doesn't deploy anywhere

I then have 2 other triggers
- A Google Cloud Build trigger which fires when a PR is merged in to our `beta` branch (basically staging environment). This builds a new docker image, deploys it, and if it is successful, moves traffic over to it
- The same for our `master` branch for our production API

I am using [supabase](https://supabase.com/), and prisma connects to it using a PSQL connection string. I have several different supabase instances for different environments, dev, test, prod.

In `prisma.schema`, I have the following datasource
```prisma
datasource db {
  provider = "postgresql"
  url      = env("SUPABASE_POSTGRES_URL")
}
```

Locally, I have a `.env` file with `SUPABASE_POSTGRES_URL=[my dev instance connection string]`. In the Google Cloud Run `beta` and `prod` containers I have a similar env var with the relevant instance connection string. 

In the builds I am running two npm scripts;
```json
    "prisma:deploy": "prisma migrate deploy",
    "prisma:generate": "prisma generate",
```

What I am confused about is when to run the deploy script to deploy any new migrations and how to manage the env var for that connection string. I have had several builds fail because the `SUPABASE_POSTGRES_URL` env var isn't available during the build so prisma can't connect to the datasource.

```
Step #0 - "Build": Step 7/9 : RUN npm run prisma:deploy
Step #0 - "Build":  ---> Running in f7a8a48a27e4
Step #0 - "Build": 
Step #0 - "Build": > loopin-api@1.0.0 prisma:deploy /usr/src/app
Step #0 - "Build": > prisma migrate deploy
Step #0 - "Build": 
Step #0 - "Build": Prisma schema loaded from prisma/schema.prisma
Step #0 - "Build": Error: Get config: Schema Parsing P1012
Step #0 - "Build": 
Step #0 - "Build": error: Environment variable not found: SUPABASE_POSTGRES_URL.
Step #0 - "Build":   -->  schema.prisma:7
Step #0 - "Build":    | 
Step #0 - "Build":  6 |   provider = "postgresql"
Step #0 - "Build":  7 |   url      = env("SUPABASE_POSTGRES_URL")
Step #0 - "Build":    | 
Step #0 - "Build": 
Step #0 - "Build": Validation Error Count: 1
Step #0 - "Build": 
Step #0 - "Build": npm ERR! code ELIFECYCLE
Step #0 - "Build": npm ERR! errno 1
Step #0 - "Build": npm ERR! loopin-api@1.0.0 prisma:deploy: `prisma migrate deploy`
Step #0 - "Build": npm ERR! Exit status 1
Step #0 - "Build": npm ERR! 
Step #0 - "Build": npm ERR! Failed at the loopin-api@1.0.0 prisma:deploy script.
Step #0 - "Build": npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
Step #0 - "Build": 
Step #0 - "Build": npm ERR! A complete log of this run can be found in:
Step #0 - "Build": npm ERR!     /root/.npm/_logs/2022-01-20T09_21_01_761Z-debug.log
Step #0 - "Build": The command '/bin/sh -c npm run prisma:deploy' returned a non-zero code: 1
Finished Step #0 - "Build"
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/docker" failed: step exited with non-zero status: 1
Step #0 - "Build": 
```

I note from the docs that I can [programmatically override a datasource url](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#programmatically-override-a-datasource-url) so I could in theory hardcode the test instance url in `prisma.schema` and then use the env var when I instantiate the client

```typescript
export const prisma = new PrismaClient({
  datasources: {
    db: {
      url: process.env.SUPABASE_POSTGRES_URL,
    },
  },
});
```

However, whilst that might fix my builds, presumably then the build would only deploy migrations to the test instance and the beta/prod instances would fall behind. 

Is my only option to add [user-defined substitutions](https://cloud.google.com/build/docs/configuring-builds/substitute-variable-values#using_user-defined_substitutions) to the three separate Cloud Build triggers and slightly messy steps to [introduce env vars into the Github actions workflow](https://stackoverflow.com/questions/60176044/how-do-i-use-an-env-file-with-github-actions)?

This seems like it would introduce quite a lot of complexity so I thought I would ask if anyone else is using this (or a similar) combination of tools and how you manage this process. If anyone has any advice, links to relevant SO questions, example repos, blog posts etc, I'd be really grateful. I've read lots of these so far, but none that have covered the situation I'm in.   

dependencies:
```json
  "@prisma/client": "^3.8.1",
  "prisma": "3.8.1",
```

**Comment by austincrim** [comment]
Hey @sauntimo 👋 , thanks for the detailed question!

How are you handling other environment variables for your per-environment builds? 

Needing to vary your builds between test, staging, prod, etc. is a pretty common pattern and most build tools have a good solution for it. I'm not familiar with Google Cloud Build, but it seems like 'user-defined substitutions' is their solution. 

Not sure if it will help your specific scenario, but we have a [CI/CD guide on our blog](https://www.prisma.io/blog/backend-prisma-typescript-orm-with-postgresql-deployment-bbba1ps7kip5).

**Reply by sauntimo** [reply]
Hey @austincrim, thanks for your comment! Currently our beta and prod API containers in Google Cloud Run each have a set of env vars, however my understanding is that those values are injected in to the node process at run time and are not available during the build. The user-defined substitutions used by Cloud Build are distinct from the env vars in Cloud Run.

I'm concerned that having env vars in so many different places (a local `.env` file, 2 containers, 3 triggers and GitHub secrets) is going to be burdensome to manage and will lead to mistakes being made. Surely there's an easier way?! It feels like I'm missing something important.

**Reply by austincrim** [reply]
@sauntimo That does sound like a lot to manage! Managing environment variables is a tough problem, there are whole platforms built just to solve that one issue (see [Doppler](https://www.doppler.com/)). GCP has its own [Secrets Manager](https://cloud.google.com/secret-manager) service you might want to look into.

I will ask, do you need to run [`prisma migrate deploy`](https://www.prisma.io/docs/reference/api-reference/command-reference#migrate-deploy) in every build pipeline? Typically, that command is only used for production or production-like environments.

**Reply by sauntimo** [reply]
Sounds like it would be good to look in to some more sophisticated secrets management - thanks for those links! Interesting point. Imagine that I have a suite of E2E API tests which I want to run against a test database instance with a known dataset, as part of my PR checks. Wouldn't I need to run the same migrations on that instance for that to work? Or would you favour manually running migrations on instances like that?

**Reply by austincrim** [reply, marked as answer]
For end-to-end tests, I would recommend that you look at [database seeding with Prisma](https://www.prisma.io/docs/guides/database/seed-database) and the [`db push`](https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push) command. 

If your test data is ephemeral, `db push` skips the migration history and prioritizes the end state of the database. You can then integrate seeding to utilize your test dataset. 

The [CI/CD article](https://www.prisma.io/blog/backend-prisma-typescript-orm-with-postgresql-deployment-bbba1ps7kip5) on our blog uses `db push` and seeding, I would recommend you take a look at that, as well as the docs linked above, to get a feel for the process.

Let me know if you have any other questions!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
